pyroclast/common/adversarial.py

```diff
@@ -0,0 +1,82 @@
+import tensorflow as tf
+import numpy as np
+
+
+def fast_gradient_sign_method(feature_fn, classify_fn, x, y, eps, norm):
+    """
+    Args:
+        feature_fn (x -> z): where x has shape [NHWCh]
+        classify_fn (z -> y): where y has shape [NCl] and z identical to feature_fn
+        x (Tensor): data, assumed to be [NHWCh]
+        y (Tensor): labels, assumed int typed
+        eps (float): norm constraint on the delta
+        norm (1, 2, or np.inf): norm choosing constraint set to optimize
+
+    Returns:
+        perturbed_data (Tensor): adversarially perturbed data to minimize \
+            feature/class outer product with shape [N,F,Cl,H,W,Ch] where each \
+            datum (H,W,Ch) is optimized against each feature/class pair (F,Cl)
+    """
+    jacobian = compute_all_gradients(feature_fn, classify_fn, x, y)
+    perturbation = linear_optimization(jacobian, eps, norm)
+    return tf.expand_dims(tf.expand_dims(x, [1]), [1]) + perturbation
+
+
+def compute_all_gradients(feature_fn, classify_fn, x, y):
+    """
+    Computes the gradient of the class logits w.r.t. input tensor
+
+    Args:
+        feature_fn (x -> z): where x has shape [NHWCh]
+        classify_fn (z -> y): where y has shape [NCl] and z identical to feature_fn
+        x (Tensor): data, assumed to be [NHWCh]
+        y (Tensor): labels, assumed int typed
+
+    Returns:
+        jacobian (Tensor): containing the gradient of the feature/class outer \
+            product with respect to the input tensor.
+    """
+    with tf.GradientTape(persistent=True,
+                         watch_accessed_variables=False) as tape:
+        tape.watch(x)
+        features = feature_fn(x)
+        fc_dot = tf.einsum('ij,ik->ijk', features, y)
+    jacobian = tape.batch_jacobian(fc_dot, x)
+    return jacobian
+
+
+def linear_optimization(jacobian, eps, norm):
+    """
```

Comment:
What does this linear optimization optimize?  Jacboian = gradients, eps = epsilon, norm = norm to minimze?